File: D_star Lite/D_star Lite_10cm/UpdateEdgeCost.m
function UpdateEdgeCost(c_old,c_uv,u,v,NewFreeFlag,NewObstacleFlag)
global g;
global c;
global rhs;
global neighbour;
global s_goal;
global xmax;
global ymax;

% New Free scanned
if c_old>c_uv
    if ~isSamePosition(u,s_goal)
        rhs(u(1),u(2))=min(rhs(u(1),u(2)),c_uv+g(v(1),v(2)));
    end
% New Obstacle scanned
elseif rhs(u(1),u(2))==c_old+g(v(1),v(2))
    if ~isSamePosition(u,s_goal)
        min_rhs=Inf;%用于对比找到min s'∈Succ(u)(c(u,s')+g(s'))
        for k=1:size(neighbour,1)%更新周边点信息
            s_prime=u+neighbour(k,:);
            if s_prime(1)<1||s_prime(1)>xmax||s_prime(2)<1||s_prime(2)>ymax
                continue
            end
            if ~isSamePosition(u,s_goal)
                if c(s_prime(1),s_prime(2))==Inf || c(u(1),u(2))==Inf
                    c_usprime=Inf;
                else
                    c_usprime=norm(u-s_prime);
                end
                if min_rhs>c_usprime+g(s_prime(1),s_prime(2))
                    min_rhs=c_usprime+g(s_prime(1),s_prime(2));
                end
            end
        end
        rhs(u(1),u(2))=min_rhs;
    end
end
UpdateVertex(u);
end

File: d_star_lite_cpp/D_start_Lite/src/UpdateEdgeCost.cpp
/* Include files */
#include <iostream>
#include <stdio.h>

#include "main.h"
#include "Initialize.h"
#include "UpdateEdgeCost.h"
#include "isSamePosition.h"
#include "rhs_cal.h"
#include "UpdateVertex.h"

/* Parameters Declarations */
static RowVector2i s_surround_updated;
static RowVector2i v;
static RowVector2i u;
static double c_old;
static double c_uv;

/* Matrix Declarations */
static RowVector3i s_updated; // (x, y, obstacle/free)

/* Function Declarations */
void UpdateEdgeCost(MatrixXi updated_region)
{
    //检测周边点情况
    for (int i = 0; i < updated_region.rows(); i++)
    {
        s_updated = updated_region.row(i).array();
        //越界跳过
        if (s_updated(0) < 1 || s_updated(0) > xmax || s_updated(1) < 1 || s_updated(1) > ymax)
        {
            continue;
        }
        //Assign current node u
        u = s_updated.head(2);
        //Scanning Map surround, Scan node for changed edge cost
        //for all directed edges(u,v) with changed edge costs
        for (int k = 0; k < 8; k++)
        {
            s_surround_updated << s_updated(0) + neighbour(k, 0), s_updated(1) + neighbour(k, 1);
            //越界跳过
            if (s_surround_updated(0) < 1 || s_surround_updated(0) > xmax || s_surround_updated(1) < 1 || s_surround_updated(1) > ymax)
            {
                continue;
            }
            //Assign surround node v
            v = s_surround_updated.head(2);

            //判断u的类型
            //new obstacle
            if (s_updated(2) == 1)
            {
                c(u(0), u(1)) = INF; //检测到obstacle，更新c
                //c_old=c(u,v)
                if (c(v(0), v(1)) == INF)//如果v是obstacle,则c_old是Inf
                {
                    c_old = INF;
                }
                else
                {
                    c_old = neighbour(k, 2);//new obstacle说明以前是free，计算c_old
                }
                c_uv = INF;//现在u是obstacle了，所以c(u,v)=Inf;
            }
            //new free
            else
            {
                c(u(0), u(1)) = 0; //检测到obstacle，更新c
                c_old = INF;//new free说明以前是obstacle, c_old一定为Inf
                if (c(v(0), v(1)) == INF)//如果v是obstacle,则c(u,v)为Inf
                {
                    c_uv = INF;
                }
                else
                {
                    c_uv = neighbour(k, 2);//new free说明以前是obstacle，计算c_old
                }
            }
            //std::cout << "c_old " << c_old << std::endl;
            //std::cout << "c_uv " << c_uv << std::endl;

            if (c_old > c_uv)
            {
                if (!isSamePosition(u, s_goal))
                {
                    //std::cout << "rhs(" << u(0) << ", " << u(1) << ") " << rhs(u(0), u(1)) << std::endl;
                    //std::cout << "g(" << v(0) << ", " << v(1) << ") " << g(v(0), v(1)) << std::endl;
                    rhs(u(0), u(1)) = fmin(rhs(u(0), u(1)), c_uv + g(v(0), v(1)));
                }
            }
            else if (rhs(u(0), u(1)) == c_old + g(v(0), v(1)))
            {
                if (!isSamePosition(u, s_goal))
                {
                    //rhs(u)=min s'∈Succ(u) (c(u,s')+g(s'))
                    rhs_cal(u);
                }
            }
            UpdateVertex(u);
            
            rhs_cal(v);
            UpdateVertex(v);
        }
    }
}